lib/events.js

```diff
@@ -351,6 +351,54 @@ EventEmitter.prototype.prependOnceListener =
       return this;
     };
 
+// An awaitable one-time event listener. Registers a once handler and
+// returns an associated promise.
+EventEmitter.prototype.when =
+    function when(type, options = { prepend: false }) {
+      if (options == null || typeof options !== 'object') {
+        // This must be lazy evaluated
+        const errors = require('internal/errors');
+        throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
+                                   'options', 'object');
+      }
+      // Do not pull these in at the top, they have to be lazy evaluated
+      const {
+        createPromise,
+        promiseReject,
+        promiseResolve
+      } = process.binding('util');
+      const promise = createPromise();
+      let resolved = false;
+      const listener = (...args) => {
+        resolved = true;
+        this.removeListener(type, listener);
+        promiseResolve(promise, { emitter: this, args });
+      }
+      Object.defineProperties(listener, {
+        name: { value: `promise for '${type}'` },
+        promise: { value: promise }
+      });
+      // When a listener is removed, and the promise has not yet been
+      // resolved, reject it with a simple reason.
+      const cancel = (eventName, removedListener) => {
+        if (!resolved && eventName === type && removedListener === listener) {
+          this.removeListener('removeListener', cancel);
+          promiseReject(promise, 'canceled');
```

Comment:
I would argue that promise rejections should always be `Error` instances. Also, this should definitely contain the information that the Promise was rejected because its listener was removed – right now, it’s quite possible that a full-blown application using this might fail and `canceled` is literally all the diagnostic information it shows.

Reply:
+1